torch/csrc/jit/codegen/cuda/type_inference.cpp

```diff
@@ -153,155 +172,124 @@ class NaiveTypePropagator {
         break;
       }
       case aten::_batch_norm_impl_index_backward: {
-        auto grad_input_type = node->input(1)->type()->cast<TensorType>();
-        TORCH_CHECK(
-            hasTypeAndDevice(grad_input_type),
-            "Type and device propagation has failed, or was not provided enough information.");
-        node->output(0)->setType(grad_input_type);
+        // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+        auto out_mask_list = constant_as<c10::List<bool>>(node->input(10));
+        TORCH_INTERNAL_ASSERT(
+            out_mask_list.has_value(),
+            "Missing output mask for batch_norm_backward");
+        std::vector<int> output_mask;
+        for (const auto value : out_mask_list->vec()) {
+          output_mask.emplace_back(static_cast<int>(value));
+        }
 
-        // TODO: double check with type promotion
-        auto mean_rstd_type = TensorType::create(
-            *grad_input_type->scalarType(),
-            *grad_input_type->device(),
-            c10::nullopt,
-            c10::nullopt);
+        auto grad_input_type = getInputTensorType(node, 1);
+        if (output_mask[0]) {
+          node->output(0)->setType(grad_input_type);
+        }
 
-        node->output(1)->setType(mean_rstd_type);
-        node->output(2)->setType(mean_rstd_type);
+        // TODO: double check with type promotion
+        if (output_mask[1]) {
+          if (auto weight_type = getInputTensorType(node, 3, true)) {
+            node->output(1)->setType(weight_type);
+          }
+        }
 
+        // TODO: Use shape information from weight tensor
+        if (output_mask[2]) {
+          auto bias_type = TensorType::create(
+              *grad_input_type->scalarType(),
+              *grad_input_type->device(),
+              c10::nullopt,
+              c10::nullopt);
+          node->output(2)->setType(bias_type);
+        }
         break;
       }
       case aten::_batch_norm_impl_index: {
-        auto out_type = node->input(0)->type()->cast<TensorType>();
-        TORCH_CHECK(
-            hasTypeAndDevice(out_type),
-            "Type and device propagation has failed, or was not provided enough information.");
+        auto out_type = getInputTensorType(node, 0);
         node->output(0)->setType(out_type);
 
-        auto mean_rstd_type = TensorType::create(
-            *out_type->scalarType(),
-            *out_type->device(),
-            c10::nullopt,
-            c10::nullopt);
-
-        node->output(1)->setType(mean_rstd_type);
-        node->output(2)->setType(mean_rstd_type);
-        // TODO: not that it matters, but mark the right type here;
-        // node->output(3)->setType(out_type->withScalarType());
-        node->output(3)->setType(out_type);
-        node->output(4)->setType(IntType::get());
+        // Assume type promotion would have consistent dtype
+        // between batch stats and running stats
+        if (auto mean_type = getInputTensorType(node, 3, true)) {
```

Comment:
Same here, also running stats output is not necessarily bundled with weight/bias. So the condition statement here is not right.